src/wallet/feebumper.cpp

```diff
@@ -43,70 +44,75 @@ int64_t CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *pWal
     return GetVirtualTransactionSize(txNew);
 }
 
-bool CFeeBumper::preconditionChecks(const CWallet *pWallet, const CWalletTx& wtx) {
+BumpFeeResult PreconditionChecks(const CWallet* pWallet, const CWalletTx& wtx, std::vector<std::string>& vErrors)
+{
     if (pWallet->HasWalletSpend(wtx.GetHash())) {
         vErrors.push_back("Transaction has descendants in the wallet");
-        currentResult = BumpFeeResult::INVALID_PARAMETER;
-        return false;
+        return BumpFeeResult::INVALID_PARAMETER;
     }
 
     {
         LOCK(mempool.cs);
         auto it_mp = mempool.mapTx.find(wtx.GetHash());
         if (it_mp != mempool.mapTx.end() && it_mp->GetCountWithDescendants() > 1) {
             vErrors.push_back("Transaction has descendants in the mempool");
-            currentResult = BumpFeeResult::INVALID_PARAMETER;
-            return false;
+            return BumpFeeResult::INVALID_PARAMETER;
         }
     }
 
     if (wtx.GetDepthInMainChain() != 0) {
         vErrors.push_back("Transaction has been mined, or is conflicted with a mined transaction");
-        currentResult = BumpFeeResult::WALLET_ERROR;
-        return false;
+        return BumpFeeResult::WALLET_ERROR;
     }
-    return true;
+    return BumpFeeResult::OK;
 }
+} // namespace
 
-CFeeBumper::CFeeBumper(const CWallet *pWallet, const uint256 txidIn, const CCoinControl& coin_control, CAmount totalFee)
-    :
-    txid(std::move(txidIn)),
-    nOldFee(0),
-    nNewFee(0)
+bool TransactionCanBeBumped(CWallet* pWallet, const uint256& txid)
 {
+    LOCK2(cs_main, pWallet->cs_wallet);
```

Comment:
It's in the original code, but is it necessary to lock `cs_main`? The same applies for other functions.

Reply:
> It's in the original code, but is it necessary to lock cs_main? The same applies for other functions.

Good catch! Fixed in 6a0131608e738f7fcf8ae9efb8ea54bf749ffe85. In the other places cs_main is required for preconditionChecks. Note that in #10973, I replace accesses to `cs_main` global throughout wallet code with `ipc_locked` locals, so it is immediately obvious where locking is and isn't required and mistakes like this will not happen in the future (see `ipc_locked` in https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/wallet/feebumper.cpp).